### PR TITLE
Keep spiked ConnectionHandler, set pool in thread

### PR DESCRIPTION
**Pull request by me**
The current functionality of `AutoReplica` is to set `ActiveRecord::Base.connection_handler` in the replica context, and reset it to the original when done. This, however, turns out to be risky.

`ActiveRecord::Base.connection_handler` is implemented using activesupport's `class_attribute`: the implementation of that is to define a method every time the variable is set, and, if the class variable is already existing, **remove the existing method and then define a new one**. This opens up to a race condition (as we observed in production): if the thread is killed in the middle of this operation (for example, by `rack_timeout`), `ActiveRecord::Base` can end up without a `connection_handler` method.

To avoid this, we propose to wrap `ConnectionHandler` once and keep it wrapped, and signal the switch to a replica pool via a thread-local variable.

### CODE REVIEW

**Comment by julik** [comment]
TL;DR - OK, but we need to look into this in a bit more detail.

I'm a bit hesitant on this _right now_ because we wanted to remove the mutex (and thread locals) around ConnectionHandler, as it seems that this protection is only required for AR3. Which we no longer use (plus - supporting only 1 Rails version back sort of makes sense). So in essence I would rollback the thread locals first, _before_ we apply this PR. @roysab what is your plan with tackling this?

Additionally I wonder - should this override (the hijack of `.connection_handler`) be applied in a Railtie then? Because otherwise one-time application might be just as unsafe as doing it on every call to replica context, if Rails materializes this method at call time?

**Comment by me** [comment]
Cool! I believe the implementation here is thread-safe, though, without the need for the mutex. In fact, I have the feeling that the current master will fail in a scenario where two threads are using different connection pools / connection params: I think the second thread will overwrite the `custom_handler`, and the first thread will be left using the new handler :/

You are very right to say that it should be applied in a Railtie. I didn't do that for simplicity, since (at least in our use case) the error is most likely to manifest itself within the replica context. But it's definitely still a gamble. Happy to work on moving the wrapping to a Railtie.

**Comment by julik** [comment]
@me what Rails version are you folks using in production? We would like to sunset AR3 support in addition to addressing the issue in this PR, in one spike.

**Comment by me** [comment]
@julik we still have AR3 in production for some apps; but we don't use threading with it, so I don't think that would affect us.

**Comment by julik** [comment]
@me will you be ok jumping on as a co-maintainer for the current major version? Then we can bump the major version, ditch AR3 support (which removes a lot of code), and you can massage the current one to do the thing you need.

**Comment by me** [comment]
@julik sure, would be happy to!

**Comment by julik** [comment]
Can you spin off a branch for 1.x support and do these changes there then? :-)

**Comment by me** [comment]
Ok! But let me understand, what's the plan? Would you like to bring these changes into master as well?

**Comment by julik** [comment]
No, the plan is that it stays on the branch for 1.x, which will still support Rails 3. Exactly because the connection_handler assignment in Rails 3 and Rails 4 differs.

**Comment by julik** [comment]
I'll close this PR so that you can work on a branch directly.

**Comment by me** [comment]
Ok!
